File: gfapy/line/group/path/validation.py
import gfapy

class Validation:
  def _validate_lists_size(self):
    n_overlaps = len(self.overlaps)
    n_segments = len(self.segment_names)
    if n_overlaps == n_segments - 1:
      # case 1: linear path
      return True
    elif n_overlaps == 1 and not self.overlaps[0]:
      # case 2: linear path, single "*" to represent overlaps which are all "*"
      return True
    elif n_overlaps == n_segments:
      # case 3: circular path
      pass
    else:
      raise gfapy.InconsistencyError(
        "Path has {} oriented segments, ".format(n_segments)+
        "but {} overlaps".format(n_overlaps))

  def _validate_record_type_specific_info(self):
    self._validate_lists_size()

File: lib/rgfa/line/path/validation.rb
module RGFA::Line::Path::Validation

  private

  def validate_lists_size!
    n_overlaps = self.overlaps.size
    n_segments = self.segment_names.size
    if n_overlaps == n_segments - 1
      # case 1: linear path
      return true
    elsif n_overlaps == 1 and self.overlaps[0].empty?
      # case 2: linear path, single "*" to represent overlaps which are all "*"
      return true
    elsif n_overlaps == n_segments
      # case 3: circular path
    else
      raise RGFA::InconsistencyError,
        "Path has #{n_segments} oriented segments, "+
        "but #{n_overlaps} overlaps"
    end
  end

  def validate_record_type_specific_info!
    validate_lists_size!
  end

end
